### PR TITLE
Addition of binary file generation in the compression method

### PR DESCRIPTION
**Pull request by jmlg97**
Thanks to this modification, cpct_pack script now can save a binary file with the compressed data. Until now, the output was an assembly file and some C headers.
This is useful, for example, to include binary compressed data to a cdt.

### CODE REVIEW

File: cpctelera/tools/scripts/cpct_pack [hunk 1]
```diff
@@ -427,6 +436,17 @@ fi
    echo "${COLOR_LIGHT_YELLOW}[cpct_pack]${COLOR_YELLOW} ${COLOR_CYAN}${OFILE}${COLOR_NORMAL} Generated."
 fi
 
+###############################################################
+###############################################################
+## Output Bin file
+##
+if (( GENERATE_B )); then
+   ## Moves the binary file generated by compressor to the output location
+   OFILE=${OUTFILE}.bin
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: Please enclose ${OUTFILE}.bin into double quotes to ensure that spaces in filenames do not cause problems
```
OFILE="${OUTFILE}.bin"
```

File: cpctelera/tools/scripts/cpct_pack [hunk 2]
```diff
@@ -427,6 +436,17 @@ fi
...
+   mv $TMPPACK ${OFILE}
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: Please, do not remove $TMPPACK by moving it. Next code in the script may use the file. Copy the contents instead.
```
   cp "$TMPPACK" "$OFILE"
```

Comment (jmlg97) [reply cpctelera/tools/scripts/cpct_pack]: True, I didn't thought that someone can add more code below this. Fixed now, thanks!

File: cpctelera/tools/scripts/cpct_pack (hunk 1 above)
Comment (jmlg97) [reply cpctelera/tools/scripts/cpct_pack]: Fine, it's done.
While i was fixing this problem, in test time I saw that other parts of code had the same problem. I fixed too.

File: cpctelera/tools/scripts/cpct_pack [hunk 3]
```diff
@@ -238,7 +247,8 @@ fi
 ## Concatenate all files in a single, temporary packfile
 echo "${COLOR_LIGHT_YELLOW}[cpct_pack]${COLOR_YELLOW} Creating single pack file (${COLOR_CYAN}${OUTFILE}${COLOR_YELLOW}) \
 with all these files: ${COLOR_CYAN}${INFILES[*]}${COLOR_NORMAL}" 
-for F in ${INFILES[*]}; do
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: What's the reason for this change? Why is the index $I needed? Previous `for`did less steps to obtain filenames in $F. Also, in this format, line 251 should be
```
F=${INFILES[$I]}
```
Has this code been tested? If there is no reason to change the code, I suggest returning to previous version.

File: cpctelera/tools/scripts/cpct_pack [hunk 4]
```diff
@@ -267,8 +277,8 @@ SAVED_BYTES=$(( UNCOMPRESSED_BYTES - COMPRESSED_BYTES ))
 if (( GENERATE_H )); then 
 
 OFILE="${OUTFILE}.h"
-cat > ${OFILE} <<TheHFile
-// File ${OFILE} generated using cpct_pack
+cat > "${OFILE}" <<TheHFile
+// File "${OFILE}" generated using cpct_pack
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: There is no need to add double quotes to the filename inside a Here Document which is just a bunch of text that will be output to a file. Spaces won't do any harm there. However, for better clarity of the sentence, as it will be a comment, using single quotes could improve readability
```
// File '${OFILE}' generated using cpct_pack
```

File: cpctelera/tools/scripts/cpct_pack [hunk 5]
```diff
@@ -325,9 +335,9 @@ fi
 ##
 if (( GENERATE_H_S )); then 
 
-OFILE=${OUTFILE}.h.s
-cat > ${OFILE} <<TheHFile
-;; File ${OFILE} generated using cpct_pack
+OFILE="${OUTFILE}.h.s"
+cat > "${OFILE}" <<TheHFile
+;; File "${OFILE}" generated using cpct_pack
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: Similar to previous comment: no need for double quotes inside a Here Document. Single quotes may be fine for readability.

File: cpctelera/tools/scripts/cpct_pack [hunk 6]
```diff
@@ -368,9 +378,9 @@ fi
 ##
 if (( GENERATE_S )); then
 
-OFILE=${OUTFILE}.s
-cat > ${OFILE} <<TheSFile
-;; File ${OFILE} generated using cpct_pack
+OFILE="${OUTFILE}.s"
+cat > "${OFILE}" <<TheSFile
+;; File "${OFILE}" generated using cpct_pack
```
Comment (lronaldo) [inline cpctelera/tools/scripts/cpct_pack]: Again, no need for double quotes inside a Here Document. Single quotes may be fine for readability.

File: cpctelera/tools/scripts/cpct_pack (hunk 3 above)
Comment (jmlg97) [reply cpctelera/tools/scripts/cpct_pack]: Of course that I tested that. As I said before, this part of the code causes problems when you intruduce a filename with spaces inside. Don't know why, but when I was testing this, I saw that "for F in ${INFILES[*]}" splits elements into different words and "cat" command below it throws file errors.

image
Execution in previous version.

image
Execution in actual version.


I recreated the error in this code
image

File: cpctelera/tools/scripts/cpct_pack (hunk 4 above)
Comment (jmlg97) [reply cpctelera/tools/scripts/cpct_pack]: Fine. I'm agree that single quotes give more clarity to the code. Changed.
However, they are needed. If they are not present, program would gives some errors specified in first picture of this message: 

> https://github.com/lronaldo/cpctelera/pull/93#discussion_r313660221

File: cpctelera/tools/scripts/cpct_pack (hunk 3 above)
Comment (lronaldo) [reply cpctelera/tools/scripts/cpct_pack]: Ok, nicely done @jmlg97. You are quiite right with this change. I dind't thought properly of it. Previous code was erroneously using * instead of @. This should also work:
```
for F in "${INFILES[@]}"; do
```
But you can perfectly leave your solution. It is pretty safe.

File: cpctelera/tools/scripts/cpct_pack (hunk 4 above)
Comment (lronaldo) [reply cpctelera/tools/scripts/cpct_pack]: Well, you are right that quotes are needed in the cat line:
```
cat > "${OFILE}" <<TheHFile
```
But they are unrequiered inside the Here Document in the next line
```
// File '${OFILE}' generated using cpct_pack
```
because that's just a bunch of literal text that is being output to a file. That bunch of text already contains spaces and adding anything with spaces won't do any harm. The only interesting piece is adding the quotes to help users differentiate the filename when they open and read these output comments.
